app/components/mappingsTable/mappingsTable.component.js

```diff
@@ -0,0 +1,224 @@
+        resetAdd();
+    }
+    function cancelAdd(){
+        resetAdd();
+    }
+    function resetAdd(){
+        vm.addView = false;
+        vm.newMapping = createNewMapping();
+        vm.decrementEditCount();
+    }
+    function createNewMapping(){
+        return {
+            conceptMapType: {},
+            conceptReferenceTerm: {
+                conceptSource: vm.sources[0]
+            },
+            termQuery: ""
+        }
+    }
+    /*adding and editing helper functions*/
+    function validMapping(map){
+        //is MapType defined? must already exist
+        if(angular.isUndefined(map.conceptMapType.uuid)){
+            openmrsNotification.error("Invalid mapping: relationship is not specified")
+            return false;
+        }
+        //fake source "search all sources..." has uuid === ""
+        if(map.conceptReferenceTerm.conceptSource.uuid === ""){
+            openmrsNotification.error("Invalid mapping: source is not specified");
+            return false;
+        }
+        //is already existing RefTerm defined?
+        var defRefTerm = angular.isDefined(map.conceptReferenceTerm.uuid);
+        //or else is new Term defined?
+        var defNewRefTerm = angular.isDefined(map.conceptReferenceTerm.conceptSource.uuid)
+                            &&angular.isDefined(map.conceptReferenceTerm.code);
+        if (defRefTerm||defNewRefTerm) return (defRefTerm||defNewRefTerm);
+        else {
+            openmrsNotification.error("Invalid Mapping: invalid reference term")
+            return false;
+        }
+    }
+    function search(mapping){
+        if(angular.isDefined(mapping.termQuery)&& mapping.termQuery.length > 1){
+            var query = {limit: 10};
+            if(angular.isDefined(mapping.conceptReferenceTerm.conceptSource.uuid)){
+                query.source = mapping.conceptReferenceTerm.conceptSource.display;
+                query.code = mapping.termQuery;
+            } else {
+                query.q = mapping.termQuery;
+            }
+            openmrsRest.listFull('conceptreferenceterm', query)
+            .then(function(response){
+                var terms = response.results;
+                mapping.suggestions = [];
+                var source = mapping.conceptReferenceTerm.conceptSource;
+                mapping.conceptReferenceTerm = {conceptSource: source};
+                for(var i =0; i<terms.length;i++){
+                    mapping.suggestions.push(terms[i]);
+                    if(mapping.termQuery === terms[i].code){
+                        mapping.conceptReferenceTerm = terms[i];
+                        mapping.conceptReferenceTerm.conceptSource = terms[i].conceptSource;
+                        break;
+                    }
+                }
+                if(angular.isUndefined(mapping.conceptReferenceTerm.uuid)) {
+                    mapping.conceptReferenceTerm.code = mapping.termQuery;
+                }
+            })
+        }
+    }
+}
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: ['mappingsTable' was used before it was defined.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741589668/issues/source?bid=3196969&fileBranchId=3354756#l224)